### PR TITLE
Fixes #24643 - Refactor Pagination to PF-React

### PR DESCRIPTION
**Pull request by glekner**
Refactoring Pagination Component to PF-React

**Issues**
The current search function removes the pagination queries from the url, while the state of the pagination component remains, this needs to be fixed at #5771 .
Changing the pagination values after searching works fine.

_TODO_
- [x] Tests

### CODE REVIEW

File: app/assets/stylesheets/base.scss [hunk 1]
```diff
@@ -337,6 +337,14 @@ table {
...
 #pagination {
   margin-top: -6px;
+
+  .form-group {
+    align-items: center;
+  }
+
+  span {
+    margin-left: 5px;
+  }
```
Comment (ohadlevy) [inline app/assets/stylesheets/base.scss]: should we move all of the CSS to the react component /webpack?

File: app/helpers/pagination_helper.rb [hunk 2]
```diff
@@ -26,4 +26,13 @@ def per_page_options(options = [5, 10, 15, 25, 50])
...
     options.uniq.sort
   end
+
+  def mount_pf_pagination(collection = nil, options = {})
+    mount_react_component('Pagination', '#pagination', {
+      :viewType => 'table',
+      :perPage => Setting[:entries_per_page],
+      :perPageOptions => per_page_options,
+      :itemCount => collection.total_entries
+      }.to_json)
```
[the remark this answers is not in the document]
Comment (ohadlevy) [reply app/helpers/pagination_helper.rb]: are you planning to change pagination on all pages? if you do, we could override the will_paginate_with_info method to update all at once

File: webpack/assets/javascripts/react_app/components/Pagination/PaginationHelper.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import URI from 'urijs';
+
+const uri = new URI(window.location.href);
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/Pagination/PaginationHelper.js]: I afraid the `uri` reviving value when the page load and won't get recalculate later.
Could you use:
```js
export const getUri => new URI(window.location.href);
```

Then you can also use it from your component.

File: webpack/assets/javascripts/react_app/components/Pagination/Pagination.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import URI from 'urijs';
+
+import { Paginator } from 'patternfly-react';
+import { changePage, changePerPage, translatePagination } from './PaginationHelper';
+import './pagination.scss';
+
+class Pagination extends React.Component {
+  render() {
+    const { data } = this.props;
+    const uri = new URI(window.location.href);
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/Pagination/Pagination.js]: You can get it from your helper `getUri`

**Review by sharvit — CHANGES_REQUESTED** [review]
Great job @glekner, some small comment

**Comment by mmoll** [comment]
@glekner please check the test failures

**Comment by boaz0** [comment]
@glekner CI fails because we need to update [this test](https://github.com/theforeman/foreman/blob/develop/test/controllers/filters_controller_test.rb#L61-L62).
I guess it's because patternfly-react pagination structure is different or the argument passed to `css_select` should be updated.

File: webpack/assets/javascripts/foreman_tools.test.js [hunk 5]
```diff
@@ -144,46 +144,38 @@ describe('updateTableTest', () => {
...
-    const PerPage = $('#per_page').val();
+    const PerPage = $('#pagination-row-dropdown').text().trim();
 
     window.location.href = 'http://localhost/?page=4';
     $('.autocomplete-input').val('test');
     $('#search-form').submit();
     expect(global.Turbolinks.visit).toHaveBeenLastCalledWith(`http://localhost/?page=1&search=test&per_page=${PerPage}`);
   });
 
-  it('should not reset search when set the per_page param', () => {
```
Comment (boaz0) [inline webpack/assets/javascripts/foreman_tools.test.js:177]: why did you remove this test?

File: test/controllers/shared/basic_rest_response_test.rb [hunk 6]
```diff
@@ -47,30 +49,14 @@ def basic_pagination_per_page_test
...
-            assert_select "option[selected='selected'][value='#{@entries_per_page + 1}']"
-            assert_select "option[value='5']"
-            assert_select "option[value='10']"
-            assert_select "option[value='15']"
-            assert_select "option[value='25']"
-            assert_select "option[value='50']"
-          end
-        end
-
-        test 'should display pagination buttons' do
```
Comment (boaz0) [inline test/controllers/shared/basic_rest_response_test.rb:67]: why is this test irrelevant?

Comment (glekner) [reply test/controllers/shared/basic_rest_response_test.rb:67]: There is no HTML of the React component here, only the ruby mounter.
I think we can count on Patternfly to test their components, especially displaying the pagination buttons..

File: webpack/assets/javascripts/foreman_tools.test.js (hunk 5 above)
Comment (glekner) [reply webpack/assets/javascripts/foreman_tools.test.js:177]: moved it to `PaginationHelper.test.js`

File: test/controllers/filters_controller_test.rb [hunk 7]
```diff
@@ -1,4 +1,5 @@
 require 'test_helper'
+require 'nokogiri'
```
Comment (ohadlevy) [inline test/controllers/filters_controller_test.rb]: probably not required for this file?

File: test/controllers/shared/basic_rest_response_test.rb [hunk 8]
```diff
@@ -1,3 +1,5 @@
+require 'nokogiri'
```
Comment (ohadlevy) [inline test/controllers/shared/basic_rest_response_test.rb]: looks like you are not using nokogiri here?

**Comment by ohadlevy** [comment]
[test katello]

**Comment by ohadlevy** [comment]
one nitpick I saw is that when hovering over the link (e.g the pagination controll) it shows the current url instead of the target url. I assume this happens because there is no href to follow ?

**Review by ohadlevy — CHANGES_REQUESTED** [review]
I didnt go over all pages, but some are broken, for example audits

**Comment by glekner** [comment]
Pagination now accepts additional classNames (e.g. Audits)

**Comment by ohadlevy** [comment]
can you verify we can actually delete https://github.com/theforeman/foreman/blob/develop/config/initializers/will_paginate.rb ?

**Comment by ohadlevy** [comment]
please also remove all unneeded methods from https://github.com/theforeman/foreman/blob/develop/app/helpers/pagination_helper.rb

**Comment by glekner** [comment]
deleted `will_paginate.rb`
removed unused method `will_paginate` from `pagination_helper.rb`

working as expected so far, lets hope tests pass.

File: app/helpers/pagination_helper.rb [hunk 9]
```diff
@@ -1,18 +1,4 @@
...
-      next_label: icon_text('angle-right', '', kind: 'fa'),
-      last_label: icon_text('angle-double-right', '', kind: 'fa'),
-      first_label: icon_text('angle-double-left', '', kind: 'fa')
-    }
-    super(collection, defaults.merge(options))
-  end
-
   def will_paginate_with_info(collection = nil, options = {})
     if collection.total_entries.zero?
       render plain: _('No entries found')
```
Comment (ohadlevy) [inline app/helpers/pagination_helper.rb]: are we using the ```per_page_options``` method?

**Comment by glekner** [comment]
@ohadlevy looks like `config/initializers/will_paginate.rb` is needed for the foreman tests to pass.

thoughts?

**Comment by ohadlevy** [comment]
does it work if you simply replace it with
require 'will_paginate/array.rb'  ?

it seems that most test failures are happening because trends seems to sort
in memory (see sort_by at
https://github.com/theforeman/foreman/blob/develop/app/controllers/trends_controller.rb#L7
)
adding the array extensions as defined at
https://www.rubydoc.info/github/mislav/will_paginate/Array

**Comment by glekner** [comment]
@ohadlevy Nice, that fixed it.

**Comment by ohadlevy** [comment]
should be probably be kept in the original initalizer file and not just in
the test (as it will make the test pass, but the ui will not work :))

**Review by ohadlevy — APPROVED** [review]

**Comment by ohadlevy** [comment]
test failures not related, thanks!

**Comment by ares** [comment]
pagination button are now clickable even when there's just one page available, I guess they should have been disabled

EDIT:

ok, I now see the difference when the button is "disabled", what's confusing is when I'm on page where there are less than "per page" records, buttons looks disabled, but clicking still looks like I'm pressing the button, minor issue though

**Comment by ohadlevy** [comment]
good catch @ares! - this is indeed a bug in pf-react, as its reproducable at [pf-react storybook](https://rawgit.com/patternfly/patternfly-react/gh-pages/index.html?knob-Item%20Count%3A=75&knob-Bordered=true&knob-Items%20End=15&knob-Selection%20Type=none&knob-Page=1&knob-Page%20Size%20Drop%20Up=true&knob-Items%20Start%3A=1&knob-Wrap%20Style=wrap&knob-Next%20button%20disabled=true&knob-Leeway=2&knob-Number%20of%20Pages=5&knob-View%20Type%3A=list&knob-Max%20Shown=5&selectedKind=patternfly-react%2FWidgets%2FPagination&selectedStory=Pagination%20row&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs).
@glekner can you please send a fix to patternfly-react? thanks!
